content/blog/2025-05-13-imperative-quantifier-elimination.md

```diff
@@ -0,0 +1,178 @@
+
+In a pure predicate logic setting, these special bounds can be "calculated" by statically unrolling a big loop that checks if each (lower bound, upper bound) pair is correct. But in Python we can just call `max` and `min`!
+
+With the least upper bound and the greatest lower bound in hand, one only needs to check if there is a value between them that is divisible by `m` (remember that `m*x` needs to satisfy all of these inequalities, not `x`) and the quantifier elimination is complete.
+
+# The Evaluation
+
+### Correctness
+
+Compiler passes (even those for class projects) should not change the behavior of the code they modify. This pass does not meet that standard for two reasons:
+
+1. As I mentioned above, this is a static transformation of untyped Python code. There is no guarantee that all of the variables in the program are integers or even that `+` does what I expect. So the best we can hope for is that it is correct _for integer inputs_.
+2. If I found all of the bugs in my implementation with testing I think that might be the first recorded instance of that happening in all of human history.
+
+That being said, I did some testing! Because this pass operates on such a simple sub-language of Python, I built a random program generator and used those programs as my test inputs. I generated around 60 programs and when one of them failed (my pass changed the program behavior) I would whittle it down to a minimal failing example and keep that as its own test.
+
+I found a lot of bugs this way, but the most frustrating one was a bug where my conjunctive normal form pass would create references to the same subtree from two different points in the AST. Then a follow-on pass would modify one of the subtrees and I would get spooky changes to portions of the AST that I was not expecting to be modifying.
+
+### Performance
+
+I chose to test the performance of my pass by testing big batches of function invocations since the running time of individual functions is very short and I was worried that overhead caused by i/o, garbage collection, or other interpreter behaviors would cause too much noise.
+
+I tested the performance improvement of my pass by taking my randomly generated test programs and running them with significantly more inputs than I tested on (I found that for testing, an input range of $[-10,10)$ was sufficient, but for performance testing I did $[-10 000,10 000)$). I also made sure to generate my inputs in a different process than the one the tested function ran in and to print out the result of each quantifier function to prevent the Python compiler from optimizing away any of the function's work. I collected my numbers using [hyperfine](https://github.com/sharkdp/hyperfine), the file numbers start at 11 because the first 10 randomly generated programs were generated with slightly more restrictive parameters, all numbers are the result of 15 runs and are reported as _mean (standard deviation)_ in milliseconds:
```

Comment:
While your approach to batching is one way to mitigate the effects of interpreter startup time, FWIW another would be to take replicated measurements _inside_ of the interpreter, à la the standard-library [timeit module](https://docs.python.org/3/library/timeit.html). [nothing to change here]

Reply:
Huh, it appears that it does turn off GC. Fascinating! https://stackoverflow.com/q/4523817